scipy/optimize/_constraints.py

```diff
@@ -0,0 +1,676 @@
+from __future__ import division, print_function, absolute_import
+import numpy as np
+import scipy.sparse as sps
+from scipy.sparse.linalg import LinearOperator
+from ._numdiff import approx_derivative
+from warnings import warn
+from ._hessian_update_strategy import HessianUpdateStrategy, SR1, BFGS
+
+
+__all__ = ['NonlinearConstraint',
+           'LinearConstraint',
+           'BoxConstraint']
+
+
+class VectorialFunction:
+    """Define methods for evaluating a vectorial function and its derivatives.
+
+    This class define a scalar function F: R^n->R^m and methods for
+    computing or approximating its first and second derivatives.
+    """
+
+    def __init__(self, fun, x0, jac='2-point', hess=BFGS(),
+                 finite_diff_options=None, sparse_jacobian=None):
```

Comment:
Just a thought: do we need `sparse_jacobian` option? Can't we simplify it? Like whatever type is returned we use it and don't force/convert anything. From the description of `sparse_jacobian` from [here](https://5226-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/optimize.minimize-trustconstr.html) I don't see its apparent value. 

If we use finite difference approximation, then it will be sparse if `sparsity` is provided, otherwise it will be dense. 

We generally should strive to remove all unnecessary or marginally useful options, there are already many things.